public/app/features/explore/TraceView/createSpanLink.tsx

```diff
@@ -72,17 +76,46 @@ export function createSpanLinkFactory({
       scopedVars = {
         ...scopedVars,
         ...scopedVarsFromSpan(span),
+        ...scopedVarsFromTags(span, traceToProfilesOptions),
       };
       // We should be here only if there are some links in the dataframe
       const fields = dataFrame.fields.filter((f) => Boolean(f.config.links?.length))!;
       try {
+        let profilesDataSourceSettings: DataSourceInstanceSettings<DataSourceJsonData> | undefined;
+        if (traceToProfilesOptions?.datasourceUid) {
+          profilesDataSourceSettings = getDatasourceSrv().getInstanceSettings(traceToProfilesOptions.datasourceUid);
+        }
+        const hasConfiguredPyroscopeDS = profilesDataSourceSettings?.type === 'grafana-pyroscope-datasource';
+        const hasPyroscopeProfile = span.tags.filter((tag) => tag.key === 'pyroscope.profile.id').length > 0;
+        const shouldCreatePyroscopeLink = hasConfiguredPyroscopeDS && hasPyroscopeProfile;
+
         let links: ExploreFieldLinkModel[] = [];
         fields.forEach((field) => {
+          let filteredLinks = field.config.links;
+          if (field.config.links) {
+            // if the span does not have a profile, we do not want to render a link to pyroscope
+            // DataLinkConfigOrigin.Datasource set on the link means that the link is defined in the datasource's dataframe
+            if (!shouldCreatePyroscopeLink && field.name === 'tags') {
+              filteredLinks = field.config.links.filter((link) =>
+                link.internal?.datasourceUid === profilesDataSourceSettings?.uid &&
+                link.origin === DataLinkConfigOrigin.Datasource
+                  ? false
+                  : true
+              );
+            }
+          }
+
           const fieldLinksForExplore = getFieldLinksForExplore({
-            field,
+            field: {
+              ...field,
+              config: {
+                ...field.config,
+                links: filteredLinks,
+              },
+            },
             rowIndex: span.dataFrameRowIndex!,
             splitOpenFn,
-            range: getTimeRangeFromSpan(span),
+            range: getTimeRangeFromSpan(span, undefined, undefined, shouldCreatePyroscopeLink),
```

Comment:
hmm yeah this is unfortunate. So what we need is a way not to render link everywhere but only if there is a specific tag? As far as I know don't we discard link if we are not able to map all the variables that we expect? At the same time we don't use the `pyroscope.profile.id` in the interpolation right?

The time range I wonder if we can use interpolation also to define time range as `from: '$span.start - 5m'` or something like that. I mean I am not sure right now if we do interpolation on time ranges and if we can mix strings and unix timestamp values like that but it could be a way to make this work.

Reply:
Great tip, I've added the pyro id to the internal link instead.

With regards to time range interpolation, I had a look but it seems like there is more to this like you said. Would like to leave as is for now.